src/components/Tabs/ControlledTabs.js

```diff
@@ -0,0 +1,211 @@
+
+
+  getCurrentPosition() {
+    const { activeTab } = this.props;
+    const { positions } = this.state;
+    return positions[activeTab];
+  }
+
+  getIdForPosition(position) {
+    const { positions } = this.state;
+    return Object.keys(positions).find(key => positions[key] === position);
+  }
+
+  getOnLinkKeyup() {
+    const linkNodes = this.linkNodes;
+
+    return (evt) => {
+      const { key } = evt;
+      const isLeftKey = key === 'ArrowLeft';
+      const isRightKey = key === 'ArrowRight';
+      if (!isLeftKey && !isRightKey) return;
+
+      const { length } = this.state;
+      const currentPosition = this.getCurrentPosition();
+
+      /* eslint-disable no-nested-ternary */
+      const nextIndex = isLeftKey
+        ? (currentPosition > 0) ? currentPosition - 1 : Math.max(length - 1, 0)
+        : (currentPosition < length - 1) ? currentPosition + 1 : 0;
+      /* eslint-enable */
+
+      const nextId = this.getIdForPosition(nextIndex);
+
+      this.onChangeTab(nextId);
+      /** Accesses the ref defined in <TabLink /> */
+      const linkNode = linkNodes.get(nextId);
+      if (linkNode && linkNode.link) { linkNode.link.focus(); }
+    };
+  }
+
+  getLinkRefSetter(id) {
+    return (c) => { this.linkNodes.set(id, c); };
+  }
+
+  getOnLinkFocus(id) {
+    return () => { this.setState({ focusedTab: id }); };
+  }
+
+  getOnBlur() {
+    return ({ currentTarget }) => {
+      /** Workaround to only fire blur when the whole navigation is unfocused */
+      setTimeout(() => {
+        if (!currentTarget.contains(document.activeElement)) {
+          this.setState({ focusedTab: false });
+        }
+      }, 0);
+    };
+  }
+
+  getOnLinkClick(id) {
+    return (evt) => {
+      evt.stopPropagation();
+      this.onChangeTab(id);
+    };
+  }
+
+  getLinkRenderer() {
+    const { focusedTab } = this.state;
+    const { activeTab, scoped } = this.props;
+
+    return (child) => {
+      const { tabTitle, id, title } = child.props;
+
+      return (
+        <TabLink
+          isActive={activeTab === id}
+          isFocused={focusedTab === id}
+          id={id}
+          onFocus={this.getOnLinkFocus(id)}
+          onKeyUp={this.getOnLinkKeyup()}
+          /* Focus is triggered before click but after mouseDown, using mouseDown to prevent a render-cycle */
+          onMouseDown={this.getOnLinkClick(id)}
+          ref={this.getLinkRefSetter(id)}
+          scoped={scoped}
+          title={tabTitle || title}
+        >
+          {title}
+        </TabLink>
+      );
+    };
+  }
+
+  getActiveTabRenderer() {
+    const { activeTab, renderInactiveTabs, scoped } = this.props;
+
+    return (child) => {
+      const { id } = child.props;
+      const isActive = id === activeTab;
+      const hideTab = !isActive && !renderInactiveTabs;
+
+      return React.cloneElement(child, {
+        'aria-labelledby': getAriaLabel(child.props.id),
+        isActive: hideTab ? null : isActive,
+        scoped
+      });
+    };
+  }
```

Comment:
These functions like `getActiveTabRenderer` or `getOnBlur` don't curry/memoize anything. Why wrap them into a getter? For `this` binding?

Reply:
They used to curry an index, I'll rewrite them.